src/uart.c

```diff
@@ -0,0 +1,405 @@
+  uart->uart_io.in.buffer_size = 0;
+  uart->uart_io.in.buffer_index = 0;
+  apbuart_set_flag(&uart->status_register, APBUART_TS);
+  apbuart_set_flag(&uart->status_register, APBUART_TE);
+}
+
+int
+uart_init_stdio(apbuart_type *uart)
+{
+  int result = 1;
+
+  if (uart != NULL) {
+    if (uart_dumbio)
+    {
+      result = 0;
+    }
+    else
+    {
+#ifdef HAVE_TERMIOS_H
+      if (uart->uart_io.in.descriptor == 0 && uart->uart_io.device.device_open)
+      {
+        tcsetattr (0, TCSANOW, &uart->uart_io.termios.io_ctrl);
+        tcflush (uart->uart_io.in.descriptor, TCIFLUSH);
+        result = 0;
+      }
+#endif
+    }
+  }
+
+  return result;
+}
+
+int
+uart_restore_stdio(apbuart_type *uart)
+{
+  int result = 1;
+
+  if (uart != NULL)
+  {
+    if (uart_dumbio)
+    {
+      result = 0;
+    }
+    else
+    {
+#ifdef HAVE_TERMIOS_H
+      if (uart->uart_io.in.descriptor == 0 && uart->uart_io.device.device_open && tty_setup)
+      {
+        tcsetattr (0, TCSANOW, &uart->uart_io.termios.io_ctrl_old);
+        result = 0;
+      }
+#endif
+    }
+  }
+
+  return result;
+}
+
+void
+apbuart_close_port (apbuart_type *uart)
+{
+  if (uart != NULL)
+  {
+    if (uart->uart_io.device.device_open && uart->uart_io.out.file != stdin)
+    {
+      fclose (uarts->uart_io.out.file); 
+    }
+  }
+}
+
+size_t
+apbuart_read_data(int read_descriptor, void *data_buffer, size_t data_size)
+{
+  size_t result = 0;
+
+  if (!(uart_dumbio || uart_nouartrx))
+  {
+    result = read (read_descriptor, data_buffer, data_size);
+  }
+
+  return result;
+}
+
+size_t
+apbuart_read_event(apbuart_type *uart)
+{
+  size_t result = 0;
+
+  if (uart != NULL)
+  {
+    if (uart->uart_io.device.device_open && apbuart_get_flag(uart->control_register, APBUART_RE))
+    {
+      result = apbuart_read_data (uart->uart_io.in.descriptor, uart->uart_io.in.buffer, 1);
+    }
+
+    if (result > 0)
+    {
+      if (apbuart_get_flag(uart->status_register, APBUART_DR))
+      {
+        apbuart_set_flag(&uart->status_register, APBUART_OV);
+      }
+      apbuart_set_flag(&uart->status_register, APBUART_DR);
+    }
+  }
+
+  return result;
+}
+
+size_t
+apbuart_fast_read_event(apbuart_type *uart)
+{
+  size_t result = 0;
+
+  if (uart->uart_io.device.device_open)
+  {
+    if (uart != NULL && !apbuart_get_flag(uart->status_register, APBUART_DR))
+    {
+      if (uart->uart_io.in.buffer_index < uart->uart_io.in.buffer_size - 1)
+      {
+        uart->uart_io.in.buffer_index++;
+        result = sizeof (uart->uart_io.in.buffer[uart->uart_io.in.buffer_index]);
+        apbuart_set_flag(&uart->status_register, APBUART_DR);
+      }
+      else
+      {
+        uart->uart_io.in.buffer_index = 0;
+        uart->uart_io.in.buffer_size = uart->uart_io.device.device_open
+          ? apbuart_read_data (uart->uart_io.in.descriptor, uart->uart_io.in.buffer, APBUART_BUFFER_SIZE)
+          : 0;
+
+        if (uart->uart_io.in.buffer_index < uart->uart_io.in.buffer_size - 1)
```

Comment:
I think that the logic of this function could be refactored to avoid code duplication.

Reply:
I thought about it and these actions are not exactly identical, partial disabling them to a separate function might not be entirely clear, so I would leave it as it is.